packages/taquito/src/prepare/prepare-provider.ts

```diff
@@ -0,0 +1,863 @@
+
+  constructor(protected context: Context) {
+    this.#counters = {};
+  }
+
+  get rpc(): RpcClientInterface {
+    return this.context.rpc;
+  }
+
+  get signer() {
+    return this.context.signer;
+  }
+
+  get estimate() {
+    return this.context.estimate;
+  }
+
+  get parser() {
+    return this.context.parser;
+  }
+
+  private async getPkh() {
+    return await this.context.signer.publicKeyHash();
+  }
+
+  private async getBlockHash() {
+    return this.context.readProvider.getBlockHash('head~2');
+  }
+
+  private async getProtocolHash() {
+    return this.context.readProvider.getNextProtocol('head');
+  }
+
+  private async getHeadCounter(pkh: string): Promise<string> {
+    return this.context.readProvider.getCounter(pkh, 'head') ?? '0';
+  }
+
+  private getFee(op: RPCOpWithFee, pkh: string, headCounter: number) {
+    if (!this.#counters[pkh] || this.#counters[pkh] < headCounter) {
+      this.#counters[pkh] = headCounter;
+    }
+    const opCounter = ++this.#counters[pkh];
+
+    return {
+      counter: `${opCounter}`,
+      fee: typeof op.fee === 'undefined' ? '0' : `${op.fee}`,
+      gas_limit: typeof op.gas_limit === 'undefined' ? '0' : `${op.gas_limit}`,
+      storage_limit: typeof op.storage_limit === 'undefined' ? '0' : `${op.storage_limit}`,
+    };
+  }
+
+  private getSource(op: RPCOpWithSource, pkh: string, source: string | undefined) {
+    return { source: typeof op.source === 'undefined' ? source || pkh : op.source };
+  }
+
+  private buildEstimates(estimate: Estimate) {
+    return {
+      fee: estimate.suggestedFeeMutez,
+      gasLimit: estimate.gasLimit,
+      storageLimit: estimate.storageLimit,
+    };
+  }
+
+  private async addRevealOperationIfNeeded(operation: RPCOperation, publicKeyHash: string) {
+    if (isOpRequireReveal(operation)) {
+      const ops: RPCOperation[] = [operation];
+      const publicKey = await this.signer.publicKey();
+      const estimateReveal = await this.estimate.reveal();
+      if (estimateReveal) {
+        const estimatedReveal = this.buildEstimates(estimateReveal);
+        ops.unshift(await createRevealOperation({ ...estimatedReveal }, publicKeyHash, publicKey));
+        return ops;
+      }
+    }
+    return operation;
+  }
+
+  private convertIntoArray(op: RPCOperation | RPCOperation[]): RPCOperation[] {
+    if (Array.isArray(op)) {
+      return [...op];
+    } else {
+      return [op];
+    }
+  }
+
+  private constructOpContents(
+    ops: RPCOperation[],
+    headCounter: number,
+    pkh: string,
+    source?: string | undefined,
+    currentVotingPeriod?: VotingPeriodBlockResult
+  ): OperationContents[] {
+    return ops.map((op: RPCOperation) => {
+      switch (op.kind) {
+        case OpKind.ACTIVATION:
+        case OpKind.DRAIN_DELEGATE:
+          return {
+            ...op,
+          };
+        case OpKind.ORIGINATION:
+          return {
+            ...op,
+            balance: typeof op.balance !== 'undefined' ? `${op.balance}` : '0',
```

Comment:
should we keep consistency with order for ternaries? here its `!==` other was `===` (if keeping the ternaries)